floss/language/go/extract.py

```diff
@@ -40,6 +43,324 @@ def extract_stackstrings(extract_stackstring_pattern, section_data, min_length)
+                continue
+
+            if not s:
+                continue
+
+            if s.encode("utf-8") != sbuf:
+                # re-encoding the string should produce the same bytes,
+                # otherwise, the string may not be the length intended.
+                continue
+
+            yield candidate
+
+
+def get_string_blob_range(pe: pefile.PE) -> Tuple[VA, VA]:
+    """
+    find the most common range of bytes between | 00 00 | pairs
+    that contains the data from a likely struct string.
+   
+    we don't expect UTF-8 data to ever contain | 00 00 |, so this
+    shouldn't be found in the string blob.
+
+    we assume that the range with the most pointers like this is
+    the string blob. this might not be the case if:
+      - there's more than one string blob
+      - there's lots of structures that look like struct String but aren't
+      - ???
+
+    these don't seem likely today, but possible.
+    """
+    image_base = pe.OPTIONAL_HEADER.ImageBase
+
+    # cache the section data so that we can avoid pefile overhead
+    section_datas: List[Tuple[VA, VA, bytes]] = []
+    for section in pe.sections:
+        if not section.IMAGE_SCN_MEM_READ:
+            continue
+
+        section_datas.append(
+            (image_base + section.VirtualAddress, 
+             image_base + section.VirtualAddress + section.SizeOfRawData,
+             section.get_data())
+        )
+
+    range_votes = collections.Counter()
+    for instance in get_struct_string_instances(pe):
+        section_start, _, section_data = next(filter(lambda s: s[0] <= instance.address < s[1], section_datas))
+
+        instance_offset = instance.address - section_start
+
+        next_null = section_data.find(b"\x00\x00", instance_offset)
+        if next_null == -1:
+            continue
+
+        prev_null = section_data.rfind(b"\x00\x00", 0, instance_offset)
+        if prev_null == -1:
+            continue
+
+        range_votes[(section_start + prev_null, section_start + next_null)] += 1
+
+    for (prev_null, next_null), count in range_votes.most_common():
+        if count == 1:
+            continue
+
+        logger.debug("range vote: 0x%x 0x%x 0x%x", prev_null + section_start, next_null + section_start, count)
+
+    most_common, count = range_votes.most_common(1)[0]
+    return most_common
+
+
+def get_string_blob_strings(pe: pefile.PE) -> Tuple[VA, str]:
+    image_base = pe.OPTIONAL_HEADER.ImageBase
+   
+    string_blob_range = get_string_blob_range(pe)
+    string_blob_start, string_blob_end = string_blob_range
+    string_blob_size = string_blob_end - string_blob_start
+    string_blob_buf = pe.get_data(string_blob_range[0] - image_base, string_blob_size)
+
+    # initialized with pointers to the beginning and end,
+    # so that when we split by pairs, we can get those bookends.
+    string_blob_pointers: List[VA] = [string_blob_range[0], string_blob_range[1]]
+
+    for instance in get_struct_string_instances(pe):
```

Comment:
the algorithm currently scans for these instances twice:
  1. to find the string blob range
  2. to extract the pointers within the range

TODO: we should extract the instances once and reuse them in both parts of the algorithm